# Workshop 07
# Systems of Linear Equations

## Inverse Matrices

In this section we consider the idea of inverse matrices and describe a common method for their construction.

As a motivation for the idea, let's again consider the system of linear equations written in the matrix form.

$$
AX = B
$$

Again, $A$ is a matrix of coefficients that are known, $B$ is a vector of known data, and $X$ is a vector that is unknown.  If $A$, $B$, and $X$ were instead only numbers, we would recognize immediately that the way to solve for $X$ is to divide both sides of the equation by $A$, so long as $A\neq 0$.  The natural question to ask about the system is *Can we define matrix division?*

The answer is *Not quite.*  We can make progress though by understanding that in the case that $A$,$B$, and $X$ are numbers, we could also find the solution by multiplying by $1/A$.  This subtle distinction is important because it means that we do not need to define division.  We only need to find the number, that when multiplied by $A$ gives 1.  This number is called the multiplicative inverse of $A$ and is written as $1/A$, so long as $A\neq 0$.

We can extend this idea to the situation where $A$, $B$, and $X$ are matrices.  In order to solve the system $AX=B$, we want to multiply by a certain matrix, that when multiplied by $A$ will give the identity matrix $I$.  This matrix is known as the **inverse matrix**, and is given the symbol $A^{-1}$.

If $A$ is a square matrix we define $A^{-1}$ (read as "A inverse") to be the matrix such that the following are true.

$$
A^{-1}A = I \hspace{3cm}AA^{-1} = I
$$

Notes about inverse matrices:

1. The matrix must be square in order for this definition to make sense.  If $A$ is not square, it is impossible for both 
$A^{-1}A$ and $AA^{-1}$ to be defined.
2. Not all matrices have inverses.  Matrices that do have inverses are called **invertible** matrices.  Matrices that do not have inverses are called **non-invertible**, or **singular**, matrices.
3. If a matrix is invertible, its inverse is unique.

Now *if we know* $A^{-1}$, we can solve the system $AX=B$ by multiplying both sides by $A^{-1}$.

$$
A^{-1}AX = A^{-1}B
$$

Then $A^{-1}AX = IX = X$, so the solution to the system is $X=A^{-1}B$.  Unfortunately, it is typically not easy to find $A^{-1}$.

### Construction of an inverse matrix

We take $C$ as an example matrix, and consider how we might build the inverse.

$$
C = \left[ \begin{array}{rrrr} 1 & 0 & 2 & -1 \\ 3 & 1 & -3 & 2 \\ 2 & 0 & 4 & 4 \\ 2 & 1 & -1 & -1 \end{array}\right]
$$

Let's think of the matrix product $CC^{-1}= I$ in terms of the columns of $C^{-1}$.  We put focus on the third column as an example, and label those unknown entries with $y_i$.  The \* entries are uknown as well, but we will ignore them for the moment.

$$
CC^{-1}=
\left[ \begin{array}{rrrr} 1 & 0 & 2 & -1 \\ 3 & 1 & -3 & 2 \\ 2 & 0 & 4 & 4 \\ 2 & 1 & -1 & -1 \end{array}\right]
\left[ \begin{array}{rrrr} * & * & y_1& * \\ * & * & y_2 & * \\ * & * & y_3 & * \\ * & * & y_4 & *  \end{array}\right]=
\left[ \begin{array}{rrrr} 1 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 \\ 0 & 0 & 1 & 0 \\ 0 & 0 & 0 & 1 \end{array}\right]=
I
$$

Recall now that $C$ multiplied by the third column of $C^{-1}$ produces the third column of $I$.  This gives us a linear system to solve for the $y_i$.

$$
\left[ \begin{array}{rrrr} 1 & 0 & 2 & -1 \\ 3 & 1 & -3 & 2 \\ 2 & 0 & 4 & 4 \\ 2 & 1 & -1 & -1 \end{array}\right]
\left[ \begin{array}{r}  y_1 \\  y_2  \\ y_3 \\ y_4  \end{array}\right]=
\left[ \begin{array}{r}  0 \\  0  \\ 1 \\ 0  \end{array}\right]
$$


In [14]:
import numpy as np

In [15]:
def RowSwap(A, k, l):
# =============================================================================
#     A is a NumPy array.  RowSwap will return duplicate array with rows
#     k and l swapped.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A
    
    B = np.copy(A).astype('float64')
        
    for j in range(n):
        temp = B[k][j]
        B[k][j] = B[l][j]
        B[l][j] = temp
        
    return B

def RowScale(A, k, scale):
# =============================================================================
#     A is a NumPy array.  RowScale will return duplicate array with the
#     entries of row k multiplied by scale.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A
    
    B = np.copy(A).astype('float64')

    for j in range(n):
        B[k][j] *= scale
        
    return B

def RowAdd(A, k, l, scale):
# =============================================================================
#     A is a numpy array.  RowAdd will return duplicate array with row
#     l modifed.  The new values will be the old values of row l added to 
#     the values of row k, multiplied by scale.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A
    
    B = np.copy(A).astype('float64')
        
    for j in range(n):
        B[l][j] += B[k][j]*scale
        
    return B

In [21]:
## Solve CY = I3
Z = np.array([[1, 0, 2, -1, 0], [3, 1, -3, 2, 0], [2, 0, 4, 4, 1], [2, 1, -1, -1, 0]])
print(Z)

[[ 1  0  2 -1  0]
 [ 3  1 -3  2  0]
 [ 2  0  4  4  1]
 [ 2  1 -1 -1  0]]


The other columns of $C^{-1}$ can be found by solving similar systems with the corresponding columns of the identity matrix.  We can then build $C^{-1}$ by assembling the columns into a single matrix, and test the result by checking the products $C^{-1}C$ and $CC^{-1}$.

In [28]:
Z = np.array([[1, 0, 2, -1, 0], [3, 1, -3, 2, 0], [2, 0, 4, 4, 1], [2, 1, -1, -1, 0]])
Z = RowAdd(Z, 0, 1, -Z[1, 0])
Z = RowAdd(Z, 0, 2, -Z[2, 0])
Z = RowAdd(Z, 0, 3, -Z[3, 0])
Z = RowScale(Z, 1, 1/Z[1, 1])
Z = RowAdd(Z, 1, 3, -Z[3, 1])
Z = RowSwap(Z, 2, 3)
Z = RowScale(Z, 2, 1/Z[2, 2])
Z = RowScale(Z, 3, 1/Z[3, 3])

print(Z)

[[ 1.          0.          2.         -1.          0.        ]
 [ 0.          1.         -9.          5.          0.        ]
 [ 0.          0.          1.         -1.          0.        ]
 [ 0.          0.          0.          1.          0.16666667]]


The following can be obtained

$$\begin{align*} 
y_1 \qquad + 2y_3 - y_4 &= 0, \\
y_2 - 9y_3 + 5y_4 &= 0, \\
y_3 - y_4 &= 0, \\
y_4 &= 0.16666667.
\end{align*}$$

We can use the function `BackSubstitution` to find $y_1, y_2, y_3, y_4$ as follows.

In [120]:
def BackSubstitution(U,B):
# =============================================================================
#     U is a NumPy array that represents an upper triangular square mxm matrix.  
#     B is a NumPy array that represents an mx1 vector     
#     BackSubstitution will return an mx1 vector that is the solution of the
#     system UX=B.
# =============================================================================
    m = U.shape[0]  # m is number of rows and columns in U
    X = np.zeros((m,1))
    
    for i in range(m-1, -1, -1):  # Calculate entries of X backward from m-1 to 0
        X[i] = B[i]
        for j in range(i+1,m):
            X[i] -= U[i][j] * X[j]
        if (U[i][i] != 0):
            X[i] /= U[i][i]
        else:
            print("Zero entry found in U pivot position",i,".")
    return X

In [121]:
# n is a number of row of Z, which is 4
(n, _) = np.shape(Z)

U = Z[:, 0:n]
print(U)
print('-' * 50)

B = Z[:, n]
print(B)
print('-' * 50)

Y = BackSubstitution(U,B)
print(Y)

[[ 1.  0.  2. -1.]
 [ 0.  1. -9.  5.]
 [ 0.  0.  1. -1.]
 [ 0.  0.  0.  1.]]
--------------------------------------------------
[ 1.      -3.       0.25    -0.33333]
--------------------------------------------------
[[ 0.83333]
 [-2.08333]
 [-0.08333]
 [-0.33333]]


We eventually obtain that

$$ 
\begin{bmatrix}
y_1 \\
y_2 \\
y_3 \\
y_4
\end{bmatrix}
=
\begin{bmatrix}
-0.16666667 \\
0.66666667 \\
0.16666667 \\
0.16666667
\end{bmatrix}
$$

We will next write a Python function to compute the inverse of a matrix.  In practice finding the inverse of a matrix is a terribly inefficient way of solving a linear system.  We have to solve $n$ systems to just to find the inverse of an $n \times n$ matrix, so it appears that it takes $n$ times the amount of work that it would to just solve the system by elimination.  Suppose however that we needed to solve a linear system $AX=B$ for *many different vectors* $B$, but the same coefficient matrix $A$.  In that case it might seem appealing to construct $A^{-1}$.

In order to keep the computation somewhat efficient, we want to avoid repeating the row operations as much as possible.  In order to construct $A^{-1}$ we need to solve the system $AX_i=Y_i$, where $Y_i$ is the $i$th column of $I$.  This will produce $X$, which is the $i$th column of $A^{-1}$.  Instead of performing elimination on each augmented matrix $[A|Y_i]$, we can augment $A$ with the entire matrix $I$ and perform the required operations on all $Y_i$ at the same time.  For example, if $A$ is a $4\times 4$ matrix, we will have the following augmented matrix.

$$
\begin{equation}
[A|I] = \left[ \begin{array}{rrrr|rrrr} 
* & * & * & * & 1 & 0 & 0 & 0 \\ 
* & * & * & * & 0 & 1 & 0 & 0 \\
* & * & * & * & 0 & 0 & 1 & 0 \\
* & * & * & * & 0 & 0 & 0 & 1 \\ 
\end{array}\right]
\end{equation}
$$

If $A$ is invertible, the $\texttt{RowReduction}$ routine from the previous section should return a matrix of the following form. 


$$
\begin{equation}
[A|I] = \left[ \begin{array}{rrrr|rrrr} 
1 & 0 & 0 & 0 & * & * & * & * \\ 
0 & 1 & 0 & 0 & * & * & * & * \\
0 & 0 & 1 & 0 & * & * & * & * \\
0 & 0 & 0 & 1 & * & * & * & * \\ 
\end{array}\right]
\end{equation}
$$


In [75]:
C = np.array([[1, 0, 2, -1], [3, 1, -3, 2], [2, 0, 4, 4], [2, 1, -1, -1]])
print(C)
print('-' * 50)

I4 = np.eye(4)
print(I4)
print('-' * 50)

CI = np.concatenate((C, I4), axis=1)
print(CI)
print('-' * 50)

[[ 1  0  2 -1]
 [ 3  1 -3  2]
 [ 2  0  4  4]
 [ 2  1 -1 -1]]
--------------------------------------------------
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
--------------------------------------------------
[[ 1.  0.  2. -1.  1.  0.  0.  0.]
 [ 3.  1. -3.  2.  0.  1.  0.  0.]
 [ 2.  0.  4.  4.  0.  0.  1.  0.]
 [ 2.  1. -1. -1.  0.  0.  0.  1.]]
--------------------------------------------------


In [94]:
CI = np.concatenate((C, I4), axis=1)
CI = RowAdd(CI, 0, 1, -CI[1,0])
CI = RowAdd(CI, 0, 2, -CI[2,0])
CI = RowAdd(CI, 0, 3, -CI[3,0])
CI = RowScale(CI, 1, 1/CI[1,1])
CI = RowAdd(CI, 1, 3, -CI[3,1])
CI = RowSwap(CI, 2, 3)
CI = RowScale(CI, 2, 1/CI[2,2])
CI = RowAdd(CI, 2, 0, -CI[0,2])
CI = RowAdd(CI, 2, 1, -CI[1,2])
CI = RowScale(CI, 3, 1/CI[3,3])
CI = RowAdd(CI, 3, 0, -CI[0,3])
CI = RowAdd(CI, 3, 1, -CI[1,3])
CI = RowAdd(CI, 3, 2, -CI[2,3])


np.set_printoptions(precision=)
print(CI)

[[ 1.       0.       0.       0.       0.83333  0.5     -0.16667 -0.5    ]
 [ 0.       1.       0.       0.      -2.08333 -1.25     0.66667  2.25   ]
 [ 0.       0.       1.       0.      -0.08333 -0.25     0.16667  0.25   ]
 [ 0.       0.       0.       1.      -0.33333  0.       0.16667  0.     ]]


In [119]:
(n, _) = np.shape(CI)
C_inverse = CI[:, n:]

print(C_inverse)

[[ 0.83333  0.5     -0.16667 -0.5    ]
 [-2.08333 -1.25     0.66667  2.25   ]
 [-0.08333 -0.25     0.16667  0.25   ]
 [-0.33333  0.       0.16667  0.     ]]


In [96]:
print(C@C_inverse)

[[ 1.00000e+00  0.00000e+00  0.00000e+00  0.00000e+00]
 [ 3.33067e-16  1.00000e+00 -5.55112e-17  0.00000e+00]
 [ 0.00000e+00  0.00000e+00  1.00000e+00  0.00000e+00]
 [ 1.11022e-16  0.00000e+00  0.00000e+00  1.00000e+00]]


If a matrix is non-invertible then the process above fails.  We have to realize that within the $\texttt{BackSubstitution}$ routine we divide by the entries along the main diagonal of the upper triangular matrix.  Recall that these entries are in the very important pivot positions. If there is a zero in at least one pivot position, then the original matrix is non-invertible.

In general we determine if a given matrix is invertible by carrying out the steps of elimination and examining the entries on the main diagonal of the corresponding upper triangular matrix.  The original matrix is invertible if and only if all of those entries are nonzoro.

### Exercises

**Exercise 1:** Solve the following system of equations using an inverse matrix.

$$
\begin{eqnarray*}
2x_1 + 3x_2 + x_3 & = & 4\\
3x_1 + 3x_2 + x_3 & = & 8\\
2x_1 + 4x_2 + x_3 & = & 5 
\end{eqnarray*}
$$

In [106]:
## Code solution here






**Exercise 2:** Let $A$ be a random $4\times 4$ matrix.  Find $A^{-1}$.

In [115]:
run = True
while run:
    A = np.random.randint(-8, 8, size=(4,4))
    da = np.linalg.det(A)
    if da != 0:
        run = False
print('A = \n', A)
print('-' * 50)

## Code solution here





A = 
 [[-1 -8 -4 -6]
 [ 7 -7  5  7]
 [ 0 -5  3 -6]
 [-1  1 -7  0]]
--------------------------------------------------


**Exercise 3:**  Solve the system $AX=B$ by finding $A^{-1}$ and computing $X=A^{-1}B$.

$$
A = \left[ \begin{array}{rrrr} 1 & 2 & -3 \\ -1 & 1 & -1  \\ 0 & -2 & 3  \end{array}\right] \quad\quad
B = \left[ \begin{array}{rrrr} 1  \\ 1 \\ 1  \end{array}\right]
$$    

In [116]:
## Code solution here




**Exercise 4:** Find a $3 \times 3 $ matrix $Y$ such that $AY = C$.

$$
A = \left[ \begin{array}{rrrr} 3 & 1 & 0 \\ 5 & 2 & 1 \\ 0 & 2 & 3\end{array}\right]\hspace{2cm}
C = \left[ \begin{array}{rrrr} 1 & 2 & 1 \\ 3 & 4 & 0 \\ 1 & 0 & 2 \end{array}\right]\hspace{2cm}
$$

In [117]:
## Code solution here





**Exercise 5:** Consider the following $ 4 \times 4 $ matrix:

$$
A = \left[ \begin{array}{rrrr} 4 & x_1 & 0 & 0 \\ 0 & x_2 & 0 & 0 \\ 0 & x_3 & 1 & 0 \\ 0 & x_4 & 0 & 3 \end{array}\right]
$$

  $(a)$ Find the condition on $x_1$, $x_2$, $x_3$ or $x_4$ for which $A^{-1}$ exists. Assuming that condition is true, find the inverse of $A$.

  $(b)$ Use Python to check if $ A^{-1}A = I $ when $x_1 = 4$, $x_2 = 1$, $x_3 = 1$, and $x_4 = 3$.

In [118]:
## Code solution here



